engine/consensus/sealing/core.go

```diff
@@ -463,46 +473,62 @@ func (c *Core) ProcessFinalizedBlock(finalizedBlockID flow.Identifier) error {
 	if err != nil {
 		c.log.Fatal().Err(err).Msgf("could not retrieve last sealed block %s", seal.BlockID)
 	}
-
-	c.log.Info().Msgf("processing finalized block %v at height %d, lastSealedHeight %d", finalizedBlockID, finalized.Height, lastSealed.Height)
-
 	c.counterLastSealedHeight.Set(lastSealed.Height)
 
-	checkEmergencySealingSpan := c.tracer.StartSpanFromParent(processFinalizedBlockSpan, trace.CONSealingCheckForEmergencySealableBlocks)
-	// check if there are stale results qualified for emergency sealing
-	err = c.checkEmergencySealing(lastSealed.Height, finalized.Height)
-	checkEmergencySealingSpan.Finish()
-	if err != nil {
-		return fmt.Errorf("could not check emergency sealing at block %v", finalizedBlockID)
-	}
+	// STEP 1: Pruning
+	// ------------------------------------------------------------------------
+	c.log.Info().Msgf("processing finalized block %v at height %d, lastSealedHeight %d", finalizedBlockID, finalized.Height, lastSealed.Height)
 
-	updateCollectorTreeSpan := c.tracer.StartSpanFromParent(processFinalizedBlockSpan, trace.CONSealingUpdateAssignmentCollectorTree)
-	// finalize forks to stop collecting approvals for orphan collectors
-	err = c.collectorTree.FinalizeForkAtLevel(finalized, lastSealed)
+	pruningSpan := c.tracer.StartSpanFromParent(processFinalizedBlockSpan, trace.CONSealingPruning)
+	err = c.collectorTree.FinalizeForkAtLevel(finalized, lastSealed) // stop collecting approvals for orphan collectors
 	if err != nil {
-		updateCollectorTreeSpan.Finish()
+		pruningSpan.Finish()
 		return fmt.Errorf("collectors tree could not finalize fork: %w", err)
 	}
 
-	pruned, err := c.collectorTree.PruneUpToHeight(lastSealed.Height)
+	pruned, err := c.collectorTree.PruneUpToHeight(lastSealed.Height) // prune AssignmentCollectorTree
 	if err != nil {
+		pruningSpan.Finish()
 		return fmt.Errorf("could not prune collectorTree tree at block %v", finalizedBlockID)
 	}
+
 	c.requestTracker.Remove(pruned...) // remove all pending items that we might have requested
-	updateCollectorTreeSpan.Finish()
 
-	err = c.sealsMempool.PruneUpToHeight(lastSealed.Height)
+	err = c.sealsMempool.PruneUpToHeight(lastSealed.Height) // prune mempool with candidate seals
 	if err != nil && !mempool.IsDecreasingPruningHeightError(err) {
 		return fmt.Errorf("could not prune seals mempool at block %v, by height: %v: %w", finalizedBlockID, lastSealed.Height, err)
```

Comment:
`pruningSpan.Finish()` before returning?
Incidentally, wouldn't extracting the successive sequences to auxiliary functions allow a `defer`-based approach ?

Reply:
👍  good idea

Reply:
I think I missed the implementation?

Reply:
I moved all the pruning logic into the auxiliary method https://github.com/onflow/flow-go/blob/21845f3f159da2e71a5e26cfc24a4a5121935afa/engine/consensus/sealing/core.go#L526-L528
which is using `defer` to close the span when exiting the method